README.md

```diff
@@ -8,21 +30,166 @@ To enable:
+
+Both standard and frame-based profiling modes are supported in the external viewer mode. See the section on frame-based profiling above for more information on this.
+
+## JSON format
+
+The JSON format is relatively simple. It contains an array of arrays, with each nested array representing a single recorded metric. The format of the array is as follows:
+
+```
+[
+    // Either 0 (ENTER CONTEXT) or 1 (EXIT CONTEXT)
+    operation: number,
+
+    // Unique number per thread. Doesn't matter what it is as long as it's unique.
+    threadId: number, 
+
+    // Current frame count. If you don't wish to use frame-based profiling, set this to -1.
+    frameId: number, 
+
+    // Timestamp of the event, in nanoseconds - this is an absolute time, not a duration
+    timestamp: number,
+
+    // The string name for the current block. Parent block names should not be included. 
+    blockName: string 
+]
+```
+
+Example JSON data:
+
+```json
+[
+  [0, 1, 0, 1609044830711675392, "DemoFunc_0"],
+  [0, 1, 0, 1609044830713163520, "DemoFunc_0_0"],
+  [0, 1, 0, 1609044830714651392, "DemoFunc_1_0"],
+  [1, 1, 0, 1609044830716141056, "DemoFunc_1_0"],
+  [1, 1, 0, 1609044830716141056, "DemoFunc_0_0"],
+  [0, 1, 0, 1609044830716141056, "DemoFunc_0_1"],
+  [0, 1, 0, 1609044830717628672, "DemoFunc_1_0"],
+  [1, 1, 0, 1609044830719115520, "DemoFunc_1_0"],
+  [1, 1, 0, 1609044830719115520, "DemoFunc_0_1"],
+  [1, 1, 0, 1609044830719115520, "DemoFunc_0"],
+  [0, 1, 1, 1609044830719115520, "DemoFunc_0"],
+  [1, 1, 1, 1609044830720603392, "DemoFunc_0"]
+]
+```
+
+Note that it is the responsibility of the application to ensure all ENTER events are balanced with a corresponding EXIT event and that the events are ordered correctly within their thread. (Cross-thread ordering does not matter - thread events can be interleaved, or each thread can have its own section of the array.) `perf_timer` will use the ordering of the ENTER and EXIT events to build call stacks for the tree views.
+
+## Binary format
+
+The binary format is similar to the JSON format, but, perhaps expectedly, is more strict. It's provided to allow for native applications to generate the output more efficiently.
+
+**Note that the binary format does not include ANY alignment padding to ensure it will work across all languages and architectures without making any assumptions about the alignment of data. Structs used to generate the metrics file should thus be packed. It is, however, assumed the output will be generated on the same system as the metrics file, so native endianness is used and it is not necessary to do any endian conversion.**
+
+The overall file format is as follows:
+
+| Field | Type | Size | Values |
+|-------|------|------|--------|
+| Magic | unsigned int32 |4 bytes | MUST contain the value 0xFA57 |
```

Comment:
Is it intentional that this field is a 32-bit uint, but the required value is a 16-bit uint?  If so, that's fine, I just got confused when I looked at it because it means half the magic number would always be zero, which is valid.

Reply:
Yeah, I'm just not fond of using 16-bit values for magic. 32-bit magic seems at least reasonably standardized.